src/lib/version.ts

```diff
@@ -0,0 +1,132 @@
+import debugFactory from 'debug';
+import semver from 'semver';
+
+const debug = debugFactory('clark:lib:version');
+
+/**
+ * Selcts the greater of two semver ranges combined with their most permissive
+ * range modifier
+ * @param left
+ * @param right
+ */
+export function select(left: string | null, right: string | null): string {
+  debug(`checking if "${left}" and "${right}" are compatible`);
+
+  // There are *much* simpler ways to write this, but typescript disagrees.
+
+  if (left === null) {
+    if (right) {
+      return right;
+    }
+    throw new Error('Cannot select a version from "null" and "null"');
+  }
+
+  if (right === null) {
+    if (left) {
+      return left;
+    }
+    throw new Error('Cannot select a version from "null" and "null"');
+  }
+
+  if (semver.intersects(left, right)) {
+    debug(`"${left}" and "${right}" are compatible`);
+
+    const leftExact = semver.clean(left.replace('^', '').replace('~', ''));
+    const rightExact = semver.clean(right.replace('^', '').replace('~', ''));
+
+    if (!leftExact) {
+      throw new Error(`"${left}" is not a valid semver`);
+    }
+
+    if (!rightExact) {
+      throw new Error(`"${right}" is not a valid semver`);
+    }
+
+    debug(`checking if "${left}" and "${right}" have the same range modifier`);
+    if (hasSameModifer(left, right)) {
+      debug(`"${left}" and "${right}" have the same range modifier`);
+      if (semver.gt(leftExact, rightExact)) {
+        return left;
+      } else {
+        return right;
+      }
+    }
+
+    debug(`"${left}" and "${right}" do not have the same range modifier`);
+
+    const modifier = extractMostPermissiveModifier(left, right);
+
+    if (semver.gt(leftExact, rightExact)) {
+      return modifier + leftExact;
+    } else {
+      return modifier + rightExact;
+    }
+  } else {
```

Comment:
When we get here, it's unclear if `semver.intersects(left, right)` returned `false` because they truly do not overlap, or because one of them is not a valid semver. Consider directly using the [`SemVer` class](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/semver/index.d.ts#L152), which throws when given an invalid string. The same could be said of calculating intersection: there is a `Range` class.